libcontainer/cgroups/fs/cpuset.go

```diff
@@ -41,7 +44,128 @@ func (s *CpusetGroup) Set(path string, cgroup *configs.Cgroup) error {
+			value, err := strconv.ParseUint(s, 10, 16)
+			if err != nil {
+				return extracted, err
+			}
+			extracted = append(extracted, uint16(value))
+		}
+	}
+
+	return extracted, nil
+}
+
 func (s *CpusetGroup) GetStats(path string, stats *cgroups.Stats) error {
+
+	var errStrings []string
+
+	if cpus, err := getCpusetStat(path, "cpuset.cpus"); err != nil {
+		errStrings = append(errStrings, err.Error())
+	} else {
+		stats.CpusetStats.Cpus = cpus
+	}
+	if cpuExclusive, err := fscommon.GetCgroupParamUint(path, "cpuset.cpu_exclusive"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.CpuExclusive = cpuExclusive
+	}
+	if mems, err := getCpusetStat(path, "cpuset.mems"); err != nil {
+		errStrings = append(errStrings, err.Error())
+	} else {
+		stats.CpusetStats.Mems = mems
+	}
+	if memHardwall, err := fscommon.GetCgroupParamUint(path, "cpuset.mem_hardwall"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.MemHardwall = memHardwall
+	}
+	if memExclusive, err := fscommon.GetCgroupParamUint(path, "cpuset.mem_exclusive"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.MemExclusive = memExclusive
+	}
+	if memoryMigrate, err := fscommon.GetCgroupParamUint(path, "cpuset.memory_migrate"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.MemoryMigrate = memoryMigrate
+	}
+	if memorySpreadPage, err := fscommon.GetCgroupParamUint(path, "cpuset.memory_spread_page"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.MemorySpreadPage = memorySpreadPage
+	}
+	if memorySpreadSlab, err := fscommon.GetCgroupParamUint(path, "cpuset.memory_spread_slab"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.MemorySpreadSlab = memorySpreadSlab
+	}
+	if memoryPressure, err := fscommon.GetCgroupParamUint(path, "cpuset.memory_pressure"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.MemoryPressure = memoryPressure
+	}
+	if schedLoadBalance, err := fscommon.GetCgroupParamUint(path, "cpuset.sched_load_balance"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.SchedLoadBalance = schedLoadBalance
+	}
+	if schedRelaxDomainLevel, err := fscommon.GetCgroupParamInt(path, "cpuset.sched_relax_domain_level"); err != nil {
+		if !os.IsNotExist(err) {
+			errStrings = append(errStrings, err.Error())
+		}
+	} else {
+		stats.CpusetStats.SchedRelaxDomainLevel = schedRelaxDomainLevel
+	}
+	if len(errStrings) != 0 {
+		return fmt.Errorf(strings.Join(errStrings, "\n"))
```

Comment:
Can we use https://github.com/hashicorp/go-multierror ?

@kolyshkin @cyphar WDYT?

Reply:
or https://github.com/uber-go/multierr

Reply:
Perhaps use `errors.New()` instead of `fmt.Errorf()` to preserve the stack

```suggestion
		return errors.New(strings.Join(errStrings, "\n"))
```

Reply:
> Can we use https://github.com/hashicorp/go-multierror ?
> or https://github.com/uber-go/multierr

Looking at the code, nothing actually "handles" these errors, other than "something failed", so from that perspective I don't think a multi-error package would be needed (to preserve the actual errors?) https://github.com/opencontainers/runc/blob/9ada2e6d4f74989e965d37bf040a4130eba13439/libcontainer/container_linux.go#L195-L197

Wondering if this function should continue trying to collect all off these, and instead just return on the first error. Currently we're updating the `stats` struct that was passed, and try to fill in as many things as possible, but the results will be thrown away if any failure occurs anyway; see https://github.com/opencontainers/runc/blob/2cf8d240075dd322b9385100c9af4b149c973391/libcontainer/cgroups/systemd/v1.go#L331-L333